src/SSTIO.cpp

```diff
@@ -0,0 +1,514 @@
+
+        std::vector<sst_cpu_map_interface_s> batch_read_data;
+        batch_read_data.reserve(max_cpus);
+        for (int i = 0; i < max_cpus; ++i) {
+            batch_read_data.emplace_back(sst_cpu_map_interface_s{static_cast<uint32_t>(i), 0});
+        }
+        auto batch_reads = ioctl_structs_from_vector<sst_cpu_map_interface_batch_s>(
+            batch_read_data);
+
+        for (auto &batch_read : batch_reads) {
+            err = ioctl(m_fd, GEOPM_IOC_SST_GET_CPU_ID, batch_read.get());
+            if (err == -1) {
+                throw Exception("SSTIOImp::SSTIOImp() failed to get CPU map",
+                                errno, __FILE__, __LINE__);
+            }
+
+            for (size_t i = 0; i < batch_read->num_entries; ++i) {
+                m_cpu_punit_core_map.emplace(batch_read->interfaces[i].cpu_index,
+                                             // Right-shift once to drop the
+                                             // hyperthread bit
+                                             batch_read->interfaces[i].punit_cpu >> 1);
+            }
+        }
+    }
+
+    int SSTIOImp::add_mbox_read(uint32_t cpu_index, uint16_t command,
+                                uint16_t subcommand, uint32_t subcommand_arg,
+                                uint32_t interface_parameter)
+    {
+        // save the stuff in the list
+        struct sst_mbox_interface_s mbox {
+            .cpu_index = cpu_index,
+            .mbox_interface_param = interface_parameter,
+            .write_value = subcommand_arg,
+            .read_value = 0,
+            .command = command,
+            .subcommand = subcommand,
+            .reserved = 0
+        };
+
+        // Stage everything in a vector for now. It will be copied to the ioctl
+        // buffer later.
+        int mbox_idx = m_mbox_read_interfaces.size();
+        m_mbox_read_interfaces.push_back(mbox);
+
+        // Multiple ioctls with different data structures are used here, along
+        // with multiple ioctl buffers. This vector indicates how a signal
+        // ID maps to a buffer, and to which offset in that buffer.
+        int idx = m_added_interfaces.size();
+        m_added_interfaces.emplace_back(MBOX, mbox_idx);
+        return idx;
+    }
+
+    int SSTIOImp::add_mbox_write(uint32_t cpu_index, uint16_t command,
+                                 uint16_t subcommand, uint32_t interface_parameter,
+                                 uint16_t read_subcommand,
+                                 uint32_t read_interface_parameter, uint32_t read_mask)
+    {
+        struct sst_mbox_interface_s mbox {
+            .cpu_index = cpu_index,
+            .mbox_interface_param = interface_parameter,
+            .write_value = 0,
+            .read_value = 0,
+            .command = command,
+            .subcommand = subcommand,
+            .reserved = 0
+        };
+        int mbox_idx = m_mbox_write_interfaces.size();
+        auto it = std::find_if(
```

Comment:
note: linear search here (same as in MSRIOGroup).  shouldn't be a problem since we have a small number of controls, and push_control is not in the hot loop.  Does it make sense to add the same check for add_mbox_read() and add_mmio*?